File: typescript/src/services/schedules-extensions.ts
import { SchedulesService as GeneratedSchedulesService } from "../generated/services/schedules.js";
import type {
  ReplaceEntryScheduleRequest,
  ScheduleEntry,
} from "../generated/services/schedules.js";
import {
  requireRecord,
  requiredWritableBoolean,
  requiredWritableString,
  writableBoolean,
  writableIdList,
  writableString,
} from "./merge-safe.js";

/** The deliberate-overwrite escape hatch named in this composite's error hints. */
const ESCAPE = "replaceEntry()";

/**
 * How the guards name this record in their messages. Spelled with a space
 * because {@link requireRecord} lowercases it into a sentence ("...where a
 * schedule entry object was expected"); "ScheduleEntry" would read as
 * "scheduleentry" there.
 */
const RECORD = "Schedule entry";

const guardOptions = { record: RECORD, escape: ESCAPE };

/**
 * Request parameters for `updateEntry`.
 *
 * **Addressedness is property presence, not value.** Every field is optional;
 * a key you leave off the object is not addressed, and a key you set is —
 * including `{ highlighted: false }`, `{ url: "" }` and
 * `{ participantIds: [] }`, each of which is a deliberate clear and reaches the
 * wire. That distinction is the whole point of the carve-out contract below,
 * so it is tested with `Object.prototype.hasOwnProperty.call`, never with a
 * truthiness test.
 *
 * The fields split in two, and the split is not cosmetic:
 *
 * - **full state** — `summary`, `startsAt`, `endsAt`, `description`, `allDay`.
 *   Always sent, seeded from the read-back when you do not address them,
 *   because BC3 rebuilds the entry from the submitted params and clears every
 *   writable field the body omits.
 * - **addressed-only** — `participantIds`, `url`, `highlighted`, `notify`.
 *   Sent only when you address them, and never seeded from the read-back. BC3
 *   preserves the first three on omission (`PRESERVED_ON_OMISSION`), so
 *   resending them is redundant at best and wrong if the read raced a
 *   concurrent change. `notify` is a directive rather than state: sending it
 *   makes BC3 recompute a drafted entry's subscriber list.
 */
export interface UpdateScheduleEntryRequest {
  /** Summary text. Omit to leave unchanged; `""` clears it. */
  summary?: string;
  /**
   * When the entry starts — a bare date (`"2026-06-05"`) for an all-day entry,
   * a timestamp otherwise. Omit to leave unchanged.
   */
  startsAt?: string;
  /** When the entry ends, in the same spelling as `startsAt`. Omit to leave unchanged. */
  endsAt?: string;
  /** Rich text description (HTML). Omit to leave unchanged; `""` clears it. */
  description?: string;
  /** Whether the entry occupies whole days. Omit to leave unchanged. */
  allDay?: boolean;
  /**
   * Replaces the entry's participants. Omit to leave them alone — BC3 preserves
   * them when the request does not address them — or pass `[]` to remove
   * everyone.
   *
   * Never resent on your behalf: the read-back's `participants` are people the
   * *reader* can see, and echoing them would rewrite the list from a stale view.
   */
  participantIds?: number[];
  /**
   * The entry's join link — a video-call URL or similar. Omit to leave it
   * alone; `""` clears it.
   *
   * Read back as `join_url`, never as `url`: an entry's `url` is its own
   * Basecamp API URL. The composite therefore never seeds this from the
   * response, which would write the API URL into the join link.
   */
  url?: string;
  /** Whether the entry is highlighted on the schedule. Omit to leave it alone; `false` removes it. */
  highlighted?: boolean;
  /**
   * Whether to notify the relevant people. A directive, not state: it is sent
   * only when you address it, and it is never read back.
   */
  notify?: boolean;
}

/**
 * A schedule entry's full writable state, handed to the `editEntry` callback.
 *
 * The five full-state members are seeded from the read-back and **always** PUT
 * back, so clearing one means setting it empty (`""`) — there is no third
 * state.
 *
 * The four carve-outs behave differently and deliberately: they are seeded for
 * *reading* (so you can inspect the current join link, highlight and
 * participants before deciding), but they reach the wire **only if you assign
 * to them**. Assignment is what marks them, not the value you assign — setting
 * `e.url = e.url` sends the join link, because intent is not recoverable from a
 * value comparison.
 */
export interface ScheduleEntryFields {
  /** Summary text. Set `""` to clear — the entry then reads back as "Untitled". */
  summary: string;
  /** Start date or timestamp, exactly as the server spelled it. */
  startsAt: string;
  /** End date or timestamp, exactly as the server spelled it. */
  endsAt: string;
  /** Rich text description (HTML). Set `""` to clear. */
  description: string;
  /** Whether the entry occupies whole days. */
  allDay: boolean;
  /** Current participants, by person ID. Assign to replace them; `[]` removes everyone. */
  participantIds: number[];
  /** The current join link (the response's `join_url`), or `""`. Assign to change it. */
  url: string;
  /** Whether the entry is currently highlighted. Assign to change it. */
  highlighted: boolean;
  /**
   * Whether to notify the relevant people. Write-only — the API reports no such
   * field, so this reads back `undefined` until you assign to it.
   */
  notify?: boolean;
}

/** The carve-out members, keyed as they appear on {@link ScheduleEntryFields}. */
const CARVE_OUTS = ["participantIds", "url", "highlighted", "notify"] as const;
type CarveOut = (typeof CARVE_OUTS)[number];

/** The subset of a replace request the carve-outs contribute. */
type CarveOutRequest = Pick<ReplaceEntryScheduleRequest, CarveOut>;

/**
 * SchedulesService with merge-safe `updateEntry` and read-modify-write
 * `editEntry` on top of the generated surface (`getEntry`, `replaceEntry`,
 * `createEntry`, `listEntries`, ...).
 *
 * `PUT /{accountId}/schedule_entries/{entryId}` is a full replace:
 * `Schedules::EntriesController#update` rebuilds the recordable from the
 * submitted params, so a sparse PUT that omits `description` erases it and one
 * that omits `all_day` turns an all-day event into a midnight-to-midnight timed
 * one. Neither is a 422 — both are a `200` that quietly clears, which is what
 * these composites exist to prevent.
 *
 * Three writable fields are exempt. BC3 seeds `participant_ids`, `url` and
 * `highlighted` from the existing recordable when the request does not address
 * them, so the SDK's job for those is the opposite one: keep them **off** the
 * wire unless the caller asked for them.
 *
 * Both composites compose the public `getEntry` and `replaceEntry`, so hooks
 * observe the two wire operations under their native identities
 * (`GetScheduleEntry` then `ReplaceScheduleEntry`) rather than a synthetic
 * composite.
 *
 * **Non-recurring entries only.** `ensure_non_recurring_event` 302-redirects
 * both `show` and `update` for a recurring entry, and the SDK follows neither:
 * the GET surfaces as an unexpected body and the PUT as a redirect. Recurrence
 * itself (`recurrence_schedule`, `recurs_until`, `time_zone_name`) is unmodeled
 * — BC3 forces all three to nil for a non-recurring entry.
 */
export class SchedulesService extends GeneratedSchedulesService {
  /**
   * Sets the given fields on a schedule entry and preserves everything else:
   * GETs the current entry, overlays the fields you addressed, and PUTs the
   * full representation back.
   *
   * A key you leave off the request is not addressed. For the five full-state
   * fields that means "keep what the server has"; for the four carve-outs it
   * means the key never appears in the body, so BC3's own preservation applies.
   * An explicitly-passed `""`, `false` or `[]` is an address and clears.
   *
   * Not atomic: there is no conditional-update signal on this endpoint, so a
   * concurrent write between the GET and PUT is overwritten — last write wins
   * for the full-state fields. The window is one round-trip. Use `replaceEntry`
   * to overwrite deliberately.
   *
   * @param entryId - The entry ID
   * @param req - Fields to set; unaddressed fields are preserved
   * @returns The updated ScheduleEntry
   * @throws {BasecampError} If the request fails, or if the read-back is malformed
   *
   * @example
   * ```ts
   * // Rename without erasing the description, times or all-day flag.
   * await client.schedules.updateEntry(123, { summary: "Team Meeting & Kickoff" });
   *
   * // Drop the join link and the highlight; participants are left alone.
   * await client.schedules.updateEntry(123, { url: "", highlighted: false });
   * ```
   */
  async updateEntry(entryId: number, req: UpdateScheduleEntryRequest): Promise<ScheduleEntry> {
    const current = await this.currentState(entryId);
    const addressed = (key: keyof UpdateScheduleEntryRequest): boolean =>
      Object.prototype.hasOwnProperty.call(req, key);

    // Full state: always sent. `??` covers the one degenerate presence case —
    // a key set to `undefined` (or, from untyped JS, `null`). Neither can ride
    // the wire, so honouring it as an address would drop the key and hand BC3
    // the clear-by-omission this composite exists to prevent; the read-back
    // value is the answer instead.
    const state: ScheduleEntryState = {
      summary: addressed("summary") ? (req.summary ?? current.summary) : current.summary,
      startsAt: addressed("startsAt") ? (req.startsAt ?? current.startsAt) : current.startsAt,
      endsAt: addressed("endsAt") ? (req.endsAt ?? current.endsAt) : current.endsAt,
      description: addressed("description")
        ? (req.description ?? current.description)
        : current.description,
      allDay: addressed("allDay") ? (req.allDay ?? current.allDay) : current.allDay,
    };

    // Addressed-only: never seeded, so presence alone decides. `false`, `""`
    // and `[]` all pass, which is the point — they are clears, and BC3 would
    // preserve rather than clear if the key went missing.
    const carveOuts: CarveOutRequest = {};
    if (addressed("participantIds")) carveOuts.participantIds = req.participantIds;
    if (addressed("url")) carveOuts.url = req.url;
    if (addressed("highlighted")) carveOuts.highlighted = req.highlighted;
    if (addressed("notify")) carveOuts.notify = req.notify;

    return this.putEntry(entryId, state, carveOuts);
  }

  /**
   * Applies a read-modify-write callback to a schedule entry: GETs the current
   * entry, hands the callback its writable state, and PUTs the whole thing
   * back. If the callback throws (or rejects), the edit aborts and nothing is
   * written.
   *
   * The five full-state fields always ride back, so clearing one means setting
   * it empty and an untouched one keeps its current value. The four carve-outs
   * (`participantIds`, `url`, `highlighted`, `notify`) are seeded so you can
   * read them, but they reach the wire only if the callback **assigns** to
   * them. Assignment is tracked by invocation, not by comparing values:
   * `e.url = e.url` is a write, because a caller who reassigns the value the
   * read returned still means "write this", and no diff can recover that
   * intent.
   *
   * Not atomic: there is no conditional-update signal on this endpoint, so a
   * concurrent write between the GET and PUT is overwritten — last write wins
   * for the full-state fields. The window is one round-trip. Use `replaceEntry`
   * to overwrite deliberately.
   *
   * @param entryId - The entry ID
   * @param fn - Callback that mutates the entry's writable fields in place
   * @returns The updated ScheduleEntry
   * @throws {BasecampError} If the request fails, or if the read-back is malformed
   *
   * @example
   * ```ts
   * await client.schedules.editEntry(123, (e) => {
   *   e.summary = `🚨 ${e.summary}`;
   *   e.description = ""; // clearing = setting empty on a full object
   *   // participants, join link and highlight are untouched, so they stay off
   *   // the wire and BC3 keeps them.
   * });
   * ```
   */
  async editEntry(
    entryId: number,
    fn: (e: ScheduleEntryFields) => void | Promise<void>
  ): Promise<ScheduleEntry> {
    const fields = await this.currentFields(entryId);

    // Setter-invocation dirty tracking. A snapshot diff would be wrong, not
    // merely different: the callback that assigns exactly what the GET returned
    // has still asked for a write, and dropping it hands the field back to
    // BC3's preserve-on-omission.
    const touched = new Set<string>();
    const view = new Proxy(fields, {
      set(target, property, value): boolean {
        if (typeof property === "string") touched.add(property);
        return Reflect.set(target, property, value);
      },
    });

    await fn(view);

    const carveOuts: CarveOutRequest = {};
    if (touched.has("participantIds")) carveOuts.participantIds = fields.participantIds;
    if (touched.has("url")) carveOuts.url = fields.url;
    if (touched.has("highlighted")) carveOuts.highlighted = fields.highlighted;
    if (touched.has("notify")) carveOuts.notify = fields.notify;

    return this.putEntry(entryId, fields, carveOuts);
  }

  /**
   * Fetches the entry and derives the full-state half of its writable state.
   *
   * Every value here is resent in the full-replace PUT, so every value is
   * validated before it is read. `?? ""` coalesces only `null` and `undefined`,
   * leaving corruption wide open: `false`, `0`, `[]`, `{}`, `42`, `true` and
   * friends would all be forwarded **verbatim** and written to the entry on a
   * call that never mentioned the field. Nothing below this rejects them —
   * `schema.d.ts` is erased at build time, so `ScheduleEntry` is a compile-time
   * claim about runtime data, which puts this composite with Python and Ruby
   * rather than with Go and Swift. See `merge-safe.ts` and #576.
   *
   * The guards differ per field because the spec models the fields
   * differently:
   *
   * - `summary` is `@required` and `Schedule::Entry#summary` is
   *   `super.presence || "Untitled"`, so blank is impossible from a healthy
   *   server and absent/null/blank is malformed.
   * - `starts_at`/`ends_at` are `@required` and `NOT NULL`. The value is a bare
   *   date for an all-day entry and a timestamp otherwise; it round-trips
   *   verbatim rather than through a `Date`, which would normalise the two
   *   spellings into one and rewrite the entry's shape.
   * - `all_day` is `@required`, `NOT NULL DEFAULT false`. It needs the boolean
   *   guard specifically because a truthiness test would read the legitimate
   *   `false` as missing — and defaulting a missing value to `false` converts
   *   an all-day event into a midnight-to-midnight timed one.
   * - `description` is optional and nullable both ways, so absent or null is a
   *   genuinely empty description.
   */
  private async currentState(entryId: number): Promise<ScheduleEntryState> {
    const current = await this.readEntry(entryId);
    return {
      summary: requiredWritableString(current, "summary", guardOptions),
      startsAt: requiredWritableString(current, "starts_at", guardOptions),
      endsAt: requiredWritableString(current, "ends_at", guardOptions),
      description: writableString(current, "description", guardOptions),
      allDay: requiredWritableBoolean(current, "all_day", guardOptions),
    };
  }

  /**
   * The full-state fields plus the carve-out seeds the `editEntry` callback can
   * read.
   *
   * `url` is seeded from the response's **`join_url`**, not its `url`: the
   * latter is the entry's own Basecamp API URL, written by
   * `recordings/_recording` before the entry partial renders. Echoing it into
   * the request's `url` would write the API URL into the join link.
   *
   * `updateEntry` deliberately does not come through here. It never seeds a
   * carve-out, so validating `participants` for it would refuse a write that
   * does not depend on them.
   */
  private async currentFields(entryId: number): Promise<ScheduleEntryFields> {
    const current = await this.readEntry(entryId);
    return {
      summary: requiredWritableString(current, "summary", guardOptions),
      startsAt: requiredWritableString(current, "starts_at", guardOptions),
      endsAt: requiredWritableString(current, "ends_at", guardOptions),
      description: writableString(current, "description", guardOptions),
      allDay: requiredWritableBoolean(current, "all_day", guardOptions),
      participantIds: writableIdList(current, "participants", guardOptions),
      url: writableString(current, "join_url", guardOptions),
      highlighted: writableBoolean(current, "highlighted", guardOptions),
    };
  }

  /**
   * GETs the entry and checks the envelope before any field is read.
   *
   * A successful GET can still return a scalar, an array or null; reading a
   * property off `null` throws a raw `TypeError` instead of the documented
   * statusless `api_error`, and reading one off an array or string silently
   * yields `undefined`, which the field guards would then take at face value.
   */
  private async readEntry(entryId: number): Promise<Record<string, unknown>> {
    return requireRecord(await this.getEntry(entryId), {
      record: RECORD,
      operation: "GetScheduleEntry",
      escape: ESCAPE,
    });
  }

  /**
   * PUTs the full writable state via `replaceEntry`, plus whichever carve-outs
   * the caller addressed.
   *
   * The five full-state fields are always named, empties included: on a
   * full-replace endpoint `""` is how a clear is expressed — never JSON null
   * (SPEC §18 body compaction), and never by omission, which would leave the
   * field to BC3's clear-by-default and read as an accident rather than an
   * intent. The carve-outs are the mirror image: an unaddressed one is absent
   * from `carveOuts`, so the generated request builder leaves it `undefined`
   * and it never reaches the wire.
   */
  private putEntry(
    entryId: number,
    state: ScheduleEntryState,
    carveOuts: CarveOutRequest
  ): Promise<ScheduleEntry> {
    return this.replaceEntry(entryId, {
      summary: state.summary,
      startsAt: state.startsAt,
      endsAt: state.endsAt,
      description: state.description,
      allDay: state.allDay,
      ...carveOuts,
    });
  }
}

/**
 * The always-sent half of a schedule entry's writable state.
 *
 * Internal: `editEntry` hands callers the wider {@link ScheduleEntryFields},
 * and `updateEntry` takes {@link UpdateScheduleEntryRequest}.
 */
interface ScheduleEntryState {
  summary: string;
  startsAt: string;
  endsAt: string;
  description: string;
  allDay: boolean;
}

File: ruby/lib/basecamp/services/schedules_extensions.rb
# frozen_string_literal: true

module Basecamp
  module Services
    # Merge-safe +update_entry+ and read-modify-write +edit_entry+ for schedule
    # entries, prepended onto the generated {SchedulesService} (see the
    # +on_load+ hook in +basecamp.rb+).
    #
    # BC3's +Schedules::EntriesController#update+ rebuilds the recordable from
    # the submitted params, so <tt>PUT /schedule_entries/{id}</tt> is a full
    # replace: a body that omits +description+ ERASES it, and one that omits
    # +summary+ erases that too — the entry then reads back as
    # <tt>"Untitled"</tt>, because +Schedule::Entry#summary+ is
    # <tt>super.presence || "Untitled"</tt>. The sparse PUT — the natural thing
    # to write — is therefore destructive on the raw endpoint, which stays
    # available as +replace_entry+.
    #
    # == Two classes of writable field
    #
    # Unlike documents and todolists, this record's writable set does not have
    # one uniform rule. It splits in two:
    #
    # [full state] +summary+, +starts_at+, +ends_at+, +description+, +all_day+.
    #              Always resent, empties included: <tt>""</tt> is how a clear is
    #              expressed on a full-replace endpoint, never JSON null and
    #              never an omission.
    # [addressed-only] +participant_ids+, +url+, +highlighted+, +notify+. Sent
    #                  *only* when the caller addressed them, and never seeded
    #                  onto the wire from the read-back.
    #
    # The first three of the addressed-only set are the operation's
    # +preservedOnOmission+ carve-out: BC3 seeds them from the existing
    # recordable when the request does not address them, so resending them is
    # redundant at best and wrong if the GET raced a concurrent change. Echoing
    # the response's +url+ would be worse than redundant — that key is the
    # entry's own Basecamp API URL, written by +recordings/_recording+ before
    # the entry partial renders, so BC3 emits the join link under the
    # non-colliding +join_url+. Writing +url+ back would store the API URL as
    # the join link. +notify+ is addressed-only for a different reason: it is a
    # directive, not state — sending it makes BC3 recompute a drafted entry's
    # subscriber list — and the read-back carries nothing to seed it from.
    #
    # An explicitly *empty* value in that second class is an address, not an
    # absence: <tt>participant_ids: []</tt> clears participants,
    # <tt>url: ""</tt> clears the join link, <tt>highlighted: false</tt> removes
    # the highlight. All three survive body compaction, which strips only nil
    # (SPEC section 18).
    #
    # == Recurring entries
    #
    # +ensure_non_recurring_event+ 302-redirects both +show+ and +update+ for a
    # recurring entry, so this route serves non-recurring entries only. The SDK
    # does not follow redirects on a PUT, and the GET's redirect lands on a body
    # this composite refuses rather than reads. Recurrence itself
    # (+recurrence_schedule+, +recurs_until+, +time_zone_name+) is unmodelled
    # here and stays unmodelled: BC3 forces all three to nil for a non-recurring
    # entry.
    #
    # Both methods compose the public +get_entry+ and +replace_entry+, so hooks
    # observe the two wire operations (+get_entry+ then +replace_entry+), not a
    # synthetic composite.
    #
    # Neither is atomic: there is no conditional-update signal on this endpoint,
    # so a concurrent write between the GET and PUT is overwritten — last write
    # wins for the whole representation. The window is one round-trip. Use
    # +replace_entry+ to overwrite deliberately.
    module SchedulesExtensions
      # The deliberate-overwrite escape hatch named in every malformed-response
      # hint raised out of this composite.
      ESCAPE_HATCH = "replace_entry"

      # The record name interpolated into {MergeSafe}'s messages.
      RECORD = "Schedule entry"

      # The writable members BC3 preserves when the request does not address
      # them, plus +notify+, which is a directive rather than state. Sent only
      # on an explicit address; never seeded from the read-back.
      CARVE_OUTS = %i[participant_ids url highlighted notify].freeze

      # How each writable member spells "cleared" on the wire.
      #
      # +compact_params+ strips nil (SPEC section 18), so a nil that reached the
      # request would silently become an omission — an address turned back into
      # an absence, which on a full-replace endpoint is exactly the defect this
      # composite exists to prevent. Every member whose type has an empty value
      # is normalised to it here. The three booleans (+all_day+, +highlighted+,
      # +notify+) have none: a boolean is true or false, so a nil assigned to
      # one is caller error and is refused rather than dropped.
      CLEARED = {
        summary: "",
        starts_at: "",
        ends_at: "",
        description: "",
        participant_ids: [].freeze,
        url: ""
      }.freeze

      # A schedule entry's writable state, yielded to the {#edit_entry} block.
      #
      # The full-state members are plain accessors: they are resent whether or
      # not the block touches them, so nothing has to be recorded. The
      # carve-outs are readable — seeded from the read-back so a block can
      # inspect the current join link, highlight and participants before
      # deciding — but each writer *records the address*, and only an addressed
      # carve-out reaches the wire.
      #
      # Dirty tracking is by setter invocation, deliberately, and not by
      # comparing the block's result against the read-back: assigning a
      # carve-out the value the GET just returned is an address like any other.
      # A diff would drop it, and the server would then hold whatever a
      # concurrent writer left there instead of the value the caller stated.
      class ScheduleEntryFields
        attr_accessor :summary, :starts_at, :ends_at, :description, :all_day
        attr_reader :participant_ids, :url, :highlighted, :notify

        def initialize(summary:, starts_at:, ends_at:, description:, all_day:,
                       participant_ids:, url:, highlighted:)
          @summary = summary
          @starts_at = starts_at
          @ends_at = ends_at
          @description = description
          @all_day = all_day
          @participant_ids = participant_ids
          @url = url
          @highlighted = highlighted
          # Nothing in the response seeds a directive.
          @notify = nil
          @addressed = {}
        end

        def participant_ids=(value)
          @addressed[:participant_ids] = true
          @participant_ids = value
        end

        def url=(value)
          @addressed[:url] = true
          @url = value
        end

        def highlighted=(value)
          @addressed[:highlighted] = true
          @highlighted = value
        end

        def notify=(value)
          @addressed[:notify] = true
          @notify = value
        end

        # Whether the caller addressed this carve-out, by assignment.
        def addressed?(name)
          @addressed.key?(name)
        end
      end

      # Sets the given fields on a schedule entry and preserves everything else:
      # GETs the current entry, overlays the explicitly-passed keyword
      # arguments, and PUTs the full representation back.
      #
      # An omitted (+nil+) argument is untouched, guaranteed. For the full-state
      # fields that means the read-back value is resent; for the addressed-only
      # fields it means the key never reaches the wire, leaving BC3 to seed it
      # from the record it already holds. An explicitly-passed <tt>""</tt>,
      # <tt>[]</tt> or +false+ is an address and is sent.
      #
      # +nil+ is an unambiguous "not addressed" for every one of these
      # arguments: none of them has a JSON null wire spelling — a clear is
      # <tt>""</tt>, <tt>[]</tt> or +false+ — and +compact_params+ strips nil
      # before serialization anyway, so no sentinel is needed to tell "passed
      # nil" from "not passed".
      #
      # Not atomic — see the module docs for the GET→PUT race, and for the
      # recurring-entry redirect. Use +replace_entry+ to overwrite deliberately.
      #
      # @param entry_id [Integer] entry id
      # @param summary [String, nil] new summary (nil = keep current)
      # @param starts_at [String, nil] new start, a date or timestamp (nil = keep current)
      # @param ends_at [String, nil] new end, a date or timestamp (nil = keep current)
      # @param description [String, nil] new description (nil = keep current, "" clears)
      # @param all_day [Boolean, nil] new all-day flag (nil = keep current)
      # @param participant_ids [Array<Integer>, nil] replaces participants (nil = leave to BC3, [] clears)
      # @param url [String, nil] new join link (nil = leave to BC3, "" clears)
      # @param highlighted [Boolean, nil] new highlight (nil = leave to BC3, false removes)
      # @param notify [Boolean, nil] notify participants (nil = do not address)
      # @return [Hash] the updated schedule entry
      def update_entry(entry_id:, summary: nil, starts_at: nil, ends_at: nil, description: nil,
                       all_day: nil, participant_ids: nil, url: nil, highlighted: nil, notify: nil)
        # Delegating to edit_entry is not a shortcut: it makes "the caller
        # addressed this" one rule with one implementation. A non-nil argument
        # invokes the same writer a block would, so the carve-outs are recorded
        # by exactly the mechanism edit_entry documents.
        edit_entry(entry_id: entry_id) do |entry|
          entry.summary = summary unless summary.nil?
          entry.starts_at = starts_at unless starts_at.nil?
          entry.ends_at = ends_at unless ends_at.nil?
          entry.description = description unless description.nil?
          entry.all_day = all_day unless all_day.nil?
          entry.participant_ids = participant_ids unless participant_ids.nil?
          entry.url = url unless url.nil?
          entry.highlighted = highlighted unless highlighted.nil?
          entry.notify = notify unless notify.nil?
        end
      end

      # Applies a read-modify-write block to a schedule entry: GETs the current
      # entry, yields its writable state ({ScheduleEntryFields}), and PUTs it
      # back. The full-state fields are resent whether or not the block touches
      # them; a carve-out is sent only if the block assigns it, even when it
      # assigns the value the read already returned. If the block raises, the
      # edit aborts and nothing is written.
      #
      # Not atomic — see the module docs for the GET→PUT race, and for the
      # recurring-entry redirect.
      #
      # @example Clear the description, leave the join link and highlight alone
      #   account.schedules.edit_entry(entry_id: 123) do |entry|
      #     entry.summary = "🚨 #{entry.summary}"
      #     entry.description = "" # clearing = setting empty on a full object
      #   end
      #
      # @example Address a carve-out
      #   account.schedules.edit_entry(entry_id: 123) do |entry|
      #     entry.url = "" if entry.url.start_with?("https://meet.example.com/")
      #   end
      #
      # @param entry_id [Integer] entry id
      # @yieldparam fields [ScheduleEntryFields] the entry's writable state, to mutate in place
      # @return [Hash] the updated schedule entry
      # @raise [ArgumentError] if no block is given
      def edit_entry(entry_id:)
        raise ArgumentError, "edit_entry requires a block" unless block_given?

        fields = fields_from_entry(get_entry(entry_id: entry_id))
        yield fields
        put_entry_fields(entry_id, fields)
      end

      private

      # Derives the writable state from a GET response.
      #
      # Every full-state value here is resent in the full-replace PUT, so every
      # one is validated before it is read. Ruby has no typed decoder between
      # the GET and this read (+get_entry+ returns a raw Hash), so the check is
      # explicit work here rather than something the layer below already did.
      # See {MergeSafe} and #576.
      #
      # The guards differ per field because the spec models the fields
      # differently:
      #
      # * +summary+, +starts_at+ and +ends_at+ are <tt>@required</tt> on the
      #   response and BC3 can never render them absent, null or blank —
      #   +Schedule::Entry#summary+ falls back to <tt>"Untitled"</tt>, and
      #   +starts_at+/+ends_at+ are NOT NULL columns every partial emits — so
      #   any of those shapes is a malformed response, not an empty value.
      # * +starts_at+/+ends_at+ are read as strings and round-tripped verbatim,
      #   never parsed: the wire value is a bare date (<tt>"2026-06-05"</tt>)
      #   for an all-day entry and a timestamp otherwise, and reformatting it
      #   would rewrite a value the caller never mentioned.
      # * +all_day+ is <tt>@required</tt> and NOT NULL DEFAULT false. It cannot
      #   be read with a truthiness test, because +false+ is the value the read
      #   most needs to admit; and defaulting a missing one to +false+ would
      #   silently convert an all-day event into a midnight-to-midnight timed
      #   one on a call that only changed the summary.
      # * +description+ is optional and nullable — the rich-text partial always
      #   sets the key but the value may be null — so absent or null is
      #   genuinely empty.
      #
      # The carve-outs are seeded for *reading* only, so a block can inspect
      # them before deciding; nothing here puts them on the wire. +url+ is
      # seeded from +join_url+, never from +url+ (see the module docs).
      # +highlighted+ is taken verbatim: it is optional, absent from the reduced
      # calendar partial +GetUpcomingSchedule+ renders, and — unlike every other
      # member — cannot reach the wire unless the caller assigns it, so there is
      # nothing to refuse a malformed value on behalf of.
      def fields_from_entry(entry)
        body = MergeSafe.require_hash(
          entry, record: RECORD, operation: "GetScheduleEntry", escape: ESCAPE_HATCH
        )
        ScheduleEntryFields.new(
          summary: required_string(body, "summary"),
          starts_at: required_string(body, "starts_at"),
          ends_at: required_string(body, "ends_at"),
          description: MergeSafe.writable_string(body, "description", record: RECORD, escape: ESCAPE_HATCH),
          all_day: MergeSafe.required_writable_boolean(body, "all_day", record: RECORD, escape: ESCAPE_HATCH),
          participant_ids: MergeSafe.writable_id_list(body, "participants", record: RECORD, escape: ESCAPE_HATCH),
          url: MergeSafe.writable_string(body, "join_url", record: RECORD, escape: ESCAPE_HATCH),
          highlighted: MergeSafe.writable_boolean(body, "highlighted", record: RECORD, escape: ESCAPE_HATCH)
        )
      end

      def required_string(body, key)
        MergeSafe.required_writable_string(body, key, record: RECORD, escape: ESCAPE_HATCH)
      end

      # PUTs the writable state via +replace_entry+.
      #
      # The five full-state members are always sent, empties included: a cleared
      # field travels as <tt>""</tt> rather than JSON null, and omitting it
      # would hand the clear back to the server's rebuild instead of stating it.
      # The carve-outs are splatted in only when addressed, so an untouched one
      # leaves no key at all and BC3 seeds it from the record it holds.
      #
      # As in Documents (#576), validating the *type* of what the caller assigns
      # is out of scope: a value the caller chose is not a value silently
      # substituted for one they asked to preserve. Normalising nil is not that
      # check — it is the addressedness contract, since a nil would be stripped
      # by +compact_params+ and turn a stated clear into an omission.
      def put_entry_fields(entry_id, fields)
        replace_entry(
          entry_id: entry_id,
          summary: caller_value(fields.summary, :summary),
          starts_at: caller_value(fields.starts_at, :starts_at),
          ends_at: caller_value(fields.ends_at, :ends_at),
          description: caller_value(fields.description, :description),
          all_day: caller_value(fields.all_day, :all_day),
          **addressed_carve_outs(fields)
        )
      end

      # The addressed carve-outs, as replace_entry keyword arguments. An
      # unaddressed one is absent from the Hash, so it is absent from the body.
      def addressed_carve_outs(fields)
        CARVE_OUTS.select { |name| fields.addressed?(name) } \
                  .to_h { |name| [ name, caller_value(fields.public_send(name), name) ] }
      end

      # Normalises a nil the caller left or assigned into that member's empty
      # spelling, and refuses it where the member has none.
      def caller_value(value, key)
        if value.nil?
          CLEARED.fetch(key) { raise nil_boolean(key) }
        else
          value
        end
      end

      def nil_boolean(key)
        UsageError.new(
          "schedule entry #{key} must be true or false, not nil",
          hint: "#{key} is a boolean with no empty value, and body compaction drops nil — " \
            "sending it would omit the field rather than state it, letting the server " \
            "decide. Assign true or false, or leave the member alone."
        )
      end
    end
  end
end
